### PR TITLE
Recover cluster after multiple pod failures

### PR DESCRIPTION
**Pull request by swoehrl-mw**
This PR implements a potential solution for cluster recovery (issue #289). 
The idea is to detect that more than one pod is missing (thus quorum potentially being broken) by checking if there are more PVCs than pods. If so the operator temporarily recreates the affected STS with `podManagementPolicy = Parallel` so that all pods are started again.

@idanl21 @prudhvigodithi  I'm not yet sure if this is a good solution, please provide your opinion on this (also if there are cases not covered) or alternative ideas you have. If you are ok with this approach I will add some tests.

I've tested two scenarios:
* All master pods down -> pods get recreated in parallel and can form a quorum
* OpenSearchCluster object is freshly created but PVCs exist (from old run) -> pods get created in parallel and can form a quorum, bootstrap pod is ignored and deleted after recovery

### CODE REVIEW

**Comment by grzeg1** [comment]
Any plans to merge this PR? We're stuck with failed cluster for the second time in a month. This time it was caused by node upgrade on AKS.

**Comment by idanl21** [comment]
Hey @swoehrl-mw, Just passed on the implementation, I didnt know the `podManagementPolicy = Parallel` but looks like it can help us to solve the problem.
did we also tested the the scenario that 2 nodes are failed an only 1 left ?
@grzeg1, Thanks for using the operator, Please -  contact me in person, i can help you with your problem (and also you can help us to test that fix). 
you are welcome to ping me in mail, idan@opster.com 
Thank !

**Comment by swoehrl-mw** [comment]
@idanl21 @prudhvigodithi I tried to add some unittests for this PR but in the end had to give up as I could not get envtest to behave like I wanted to. We'll have to cover this with functional tests in the future.
I added an option to disable the recovery so if we didn't cover an edge case the users can prevent it from doing stupid stuff.
Please review and also test in your environments.

Tests I did:
* Delete all pods -> pods get recreated in parallel
* Delete 3 of 4 pods -> pods get recreated in parallel
* Delete 2 of 3 pods -> normal one-by-one recovery works (2 can form a quorum)
* Delete cluster, keep PVCs, recreate cluster -> pods get created in parallel

File: docs/userguide/main.md [hunk 1]
```diff
@@ -682,6 +682,13 @@ Internally you should use self-signed certificates (you can let the operator gen
 
 The operator contains several features that automate management tasks that might be needed during the cluster lifecycle. The different available options are documented here.
 
+### Cluster recovery
+
+This operator automatically handles common failure scenarios and restarts crashed pods, normally this is done in a one-by-one fashion to maintain quorum and cluster stability.
+In case the operator detects several crashed or missing pods (for a nodepool) at the same time it will switch into a special recovery mode and start all pods at once and allow the cluster to form a new quorum. This parallel recovery mode is currently experimental and only works with PVC-backed storage. If you encounter problems with it, you can disable it by redeploying the operator and adding `manager.parallelRecoveryEnabled: false` to your `values.yaml`. Please also report any problems by opening an issue in the operator github project.
+
```
Comment (idanl21) [inline docs/userguide/main.md:689]: Maybe add also how we are doing it ?
And, do we return the podManagementPolicy to his original value after the cluster is back to normal ?

File: opensearch-operator/controllers/tls_test.go [hunk 2]
```diff
@@ -49,9 +48,10 @@ var _ = Describe("TLS Reconciler", func() {
 				}},
 				NodePools: []opsterv1.NodePool{
 					{
-						Component: "masters",
-						Replicas:  3,
-						Roles:     []string{"master", "data"},
+						Component:   "masters",
+						Replicas:    3,
+						Roles:       []string{"master", "data"},
+						Persistence: &opsterv1.PersistenceConfig{PersistenceSource: opsterv1.PersistenceSource{EmptyDir: &corev1.EmptyDirVolumeSource{}}},
```
Comment (idanl21) [inline opensearch-operator/controllers/tls_test.go:54]: thats related to the PR?

File: opensearch-operator/pkg/helpers/helpers.go [hunk 3]
```diff
@@ -159,3 +164,70 @@ func DiffSlice(leftSlice, rightSlice []string) []string {
...
+	selector := labels.NewSelector()
+	selector = selector.Add(*clusterReq, *componentReq)
+	list := corev1.PersistentVolumeClaimList{}
+	if err := k8sClient.List(ctx, &list, &client.ListOptions{LabelSelector: selector}); err != nil {
+		return 0, err
+	}
+	return len(list.Items), nil
+}
+
+// Delete a STS with cascade=orphan and wait until it is actually deleted from the kubernetes API
```
Comment (idanl21) [inline opensearch-operator/pkg/helpers/helpers.go:187]: what is STS with cascade ?

File: opensearch-operator/pkg/helpers/helpers.go [hunk 4]
```diff
@@ -159,3 +164,70 @@ func DiffSlice(leftSlice, rightSlice []string) []string {
...
+func WaitForSTSStatus(ctx context.Context, k8sClient client.Client, obj *appsv1.StatefulSet) (*appsv1.StatefulSet, error) {
+	for i := 1; i <= 10; i++ {
+		existing := appsv1.StatefulSet{}
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(obj), &existing)
+		if err == nil {
+			if existing.Status.CurrentRevision != "" {
+				return &existing, nil
+			}
+		}
+		time.Sleep(time.Second * 3)
```
Comment (idanl21) [inline opensearch-operator/pkg/helpers/helpers.go:230]: I think we need to declare the 30 sec as a parameter

File: opensearch-operator/pkg/reconcilers/cluster.go [hunk 5]
```diff
@@ -149,15 +149,69 @@ func (r *ClusterReconciler) reconcileNodeStatefulSet(nodePool opsterv1.NodePool,
 	// Fix selector.matchLabels (issue #311), need to recreate the STS for it as spec.selector is immutable
 	if _, exists := existing.Spec.Selector.MatchLabels["opensearch.role"]; exists {
 		r.logger.Info("deleting statefulset while orphaning pods to fix labels " + existing.Name)
-		opts := client.DeleteOptions{}
-		client.PropagationPolicy(metav1.DeletePropagationOrphan).ApplyToDelete(&opts)
-		if err := r.Delete(r.ctx, existing, &opts); err != nil {
-			r.logger.Info("failed to delete statefulset" + existing.Name)
+		if err := helpers.WaitForSTSDelete(r.ctx, r.Client, existing); err != nil {
+			r.logger.Error(err, "Failed to delete Statefulset for nodePool "+nodePool.Component)
```
Comment (idanl21) [inline opensearch-operator/pkg/reconcilers/cluster.go:153]: How do you detect un-healthy pod? is it a pod that is not on running status? or 0/1 is also un-healthy?

**Comment by grzeg1** [comment]
> Hey @swoehrl-mw, Just passed on the implementation, I didnt know the `podManagementPolicy = Parallel` but looks like it can help us to solve the problem. did we also tested the the scenario that 2 nodes are failed an only 1 left ? @grzeg1, Thanks for using the operator, Please - contact me in person, i can help you with your problem (and also you can help us to test that fix). you are welcome to ping me in mail, [idan@opster.com](mailto:idan@opster.com) Thank !

@idanl21 we had to back from using the operator and return to manually-provisioned cluster. But we're ready to go back to testing.

File: docs/userguide/main.md (hunk 1 above)
Comment (swoehrl-mw) [reply docs/userguide/main.md:689]: I've extended the explanation a bit. 
And yes, as soon as all pods are running the podManagemtnPolicy is put back.

File: opensearch-operator/controllers/tls_test.go (hunk 2 above)
Comment (swoehrl-mw) [reply opensearch-operator/controllers/tls_test.go:54]: The test was running into problems because no configured persistence means PVC and that meant the recovery code tried to kick in during the test. By changing the persistence it is ignored.

File: opensearch-operator/pkg/helpers/constants.go [hunk 6]
```diff
@@ -19,3 +25,17 @@ func ClusterDnsBase() string {
 
 	return env
 }
+
+func ParallelRecoveryMode() bool {
+	env, found := os.LookupEnv(ParallelRecoveryEnabled)
```
[the remark this answers is not in the document]
Comment (swoehrl-mw) [reply opensearch-operator/pkg/helpers/constants.go:30]: go standard library function to get the value of an environment variable.

File: opensearch-operator/pkg/helpers/helpers.go (hunk 3 above)
Comment (swoehrl-mw) [reply opensearch-operator/pkg/helpers/helpers.go:192]: `cascade=orphan` means kubernetes will not delete the pods of the statefulset. That way any remaining pods keep running and are picked up again as soon as we recreate the STS.

File: opensearch-operator/pkg/helpers/helpers.go (hunk 4 above)
Comment (swoehrl-mw) [reply opensearch-operator/pkg/helpers/helpers.go]: I've moved that to a constant. I'd like to avoid having it as a configurable parameter (just adds complexity). Even if in an environment the waittime is too short the operator will just log an error and requeue and try again, so no harm.

File: opensearch-operator/pkg/reconcilers/cluster.go (hunk 5 above)
Comment (swoehrl-mw) [reply opensearch-operator/pkg/reconcilers/cluster.go:153]: The recovery process does not care for unhealthy pods, it just cares about the pods existing and assumes they will either start up and become healthy or kubernetes will take care of it.

**Comment by idanl21** [comment]
> 

Hey @grzeg1, Im glad that you guys wanna keep work and progress with the Operator, i believe that fix will be merged to main in few days :) 
If you have additional questions, please let me know send me an email (idan@opster.com)

**Review by idanl21 — APPROVED** [review]
Ready to merge form my point of view :)

**Comment by grzeg1** [comment]
> > 
> 
> Hey @grzeg1, Im glad that you guys wanna keep work and progress with the Operator, i believe that fix will be merged to main in few days :) If you have additional questions, please let me know send me an email ([idan@opster.com](mailto:idan@opster.com))

Just to let you know: we've been using the merged fix for a few days and tried to make the cluster break. We did not succeed - the cluster always recovered. Thumbs up!
